docs/spec/overview.md

```diff
@@ -1,70 +1,127 @@
+services or earlier computation) for further event processing and persistence.
 
-* _Feeds_ bridge the event source into the eventing framework.
+The following is a detailed specification for the `eventing.knative.dev` API
+surface, which provides control mechanism for Knative eventing.
 
-* _Flows_ abstract the path of an event from a source takes to reach an event
-  consumer.
+# Resource Types
 
-Typically, _Feeds_ perform out-of-cluster provisioning, while _Channels_ are
-within-cluster management of event delivery. _Flows_ is a higher order wrapper
-used to connect event producers directly with event consumers leveraging 
-_Feeds_ and _Channels_.
+The API defines and provides a complete implementation for
+[Subscription](spec.md#kind-subscription), and abstract resource definitions
+for [Sources](spec.md#kind-source), [Channels](spec.md#kind-channel), and
+[Providers](spec.md#kind-provisioner) which may be fulfilled by multiple
+backing implementations (much like the Kubernetes Ingress resource).
 
-# Resource Types
+- A **Subscription** describes the transformation of an event and optional
+  forwarding of a returned event.
+
+- A **Source** allows an incoming events from an external system to be
+  _Subscribable_. A _Subscription_ is used to connect these events to
+  subsequent processing steps.
+
+- A **Channel** provides event persistance and fanout of events from a
+  well-known input address to multiple outputs described by _Subscriptions_.
+
+<!-- This image is sourced from https://drive.google.com/open?id=10mmXzDb8S_4_ZG_hcBr7s4HPISyBqcqeJLTXLwkilRc -->
+
+![Resource Types Overview](images/resource-types-overview.svg)
+
+- **Provisioners** implement strategies for realizing backing resources for
+  different implementations of _Sources_ and _Channels_ currently active in the
+  eventing system.
+
+<!-- This image is sourced from https://drive.google.com/open?id=1o_0Xh5VjwpQ7Px08h_Q4qnaOdMjt4yCEPixRFwJQjh8 -->
 
-The primary resources in the Knative Eventing _Feeds_ API are EventSource,
-EventType and Feed:
+![Resource Types Provisioners](images/resource-types-provisioner.svg)
 
-* **EventSource**, an archetype of an event producer.
+With extendibility and compostability as a goal of Knative Eventing, the
+eventing API defines several resources that can be reduced down to a well
+understood contracts. These eventing resource interfaces may be fulfilled by
+other Kubernetes objects and then composed in the same way as the concreate
+objects. The interfaces are ([Sinkable](interfaces.md#sinkable),
+[Subscribable](interfaces.md#subscribable),
+[Channelable](interfaces.md#channelable),
+[Targetable](interfaces.md#targetable)). For more details, see
+[Interface Contracts](interfaces.md).
 
-* **EventType**, the schema for an event. 
+## Subscription
 
-* **Feed**, the association between the output of an event producer to the
-  input of an event consumer.
+**Subscriptions** describe a flow of events from one event producer or
```

Comment:
* Each `Subscription` is basically tied to two "components" (e.g. event source sends events to some target).
* Now that _target_ could process the data, act as a _source_ and send it to a different target
* and so on...

All of these `Subscritptions` may represent some "Workflow".  Are there plans to cover this too? (Perhaps the the via the "CNCF Serverless WG's `workflow` spec)  Just wondering - I do *NOT* think that's in scope for 0.1 - but just curious

Reply:
I think there are interesting questions about how to chain Targetables together.

For the scope of 0.1, the only option we support is sinking the result of one Targetable into a Channel, and then creating a Subscription from that Channel which calls the next Targetable.

Reply:
Hm. This is interesting. Let me shop the idea around a bit and get back to you.

Reply:
> Now that target could process the data, act as a source and send it to a different target

@matzew If I understand what you're proposing, this is a common anti-pattern in serverless. Each nested call ends up causing the original call to block preventing the component from scaling down once it had completed its work. 

![nested calls](https://user-images.githubusercontent.com/302992/46741297-b9c78980-cc72-11e8-9d66-6f8bf5355a5c.png)

The deeper the chain, the more components are blocked waiting for a reply, doing no real work, but generating billable time.

It's possible to avoid the blocking, but then you lose any ability to detect and recover from errors.

![linear calls](https://user-images.githubusercontent.com/302992/46741529-36f2fe80-cc73-11e8-81a3-4990d2087a07.png)

The current Subscription model creates a chain or responsibility where channels mediate the events creating both error boundaries and the ability to retry after failures.

![chain of responsibility-2](https://user-images.githubusercontent.com/302992/46741628-70c40500-cc73-11e8-9768-e858763c7767.png)

Treat any references to "Call Pipe" as a "Channel" these diagrams were created against a much, much earlier proposal.

Reply:
> Now that target could process the data, act as a source and send it to a different target

@scothis  one more comment/thought.

So, the `target` (aka function, aka `ksvc`) receives data. But it would not do a direct (http) invocation. I think the function here could act as a _true_ event-source: Delivering an _async_ event, after it (the `target`) did process the data.

Now, this _event-source_ (implemented as a function) can have subscriptions again ... 

@n3wscott @evankanderson  any thoughts ?

Reply:
Ok... I think my comments are _irrelevant_ for *this* PR - I will open an issue for discussions